### ISSUE TITLE
Two-pass entity extraction: GLiNER for entities, targeted LLM for relationships

### ISSUE THREAD

**Issue by zonk1024** [ourochronos/covalence#32]
## Context

Related to #11 (fine-tune relationship model) and #12 (emergent ontology). This is the extraction pipeline redesign.

## Current State

Single-pass LLM extraction in `llm_extractor.rs`: sends chunk text to OpenAI, asks it to extract entities AND relationships in one shot. The prompt is fully open-ended:

```
You are an entity and relationship extractor. Given a text passage, extract all notable entities and relationships.
```

This is expensive, hallucination-prone, and unconstrained.

**GLiNER extractor already exists** in `gliner_extractor.rs` — full HTTP sidecar client with entity type detection. It just needs a sidecar to talk to.

### Pass 1: GLiNER (local, fast, free)
- Extract entities with types and spans
- ~10-50ms per chunk, zero API cost
- High precision (span-based, grounded in text)
- If zero entities found → skip Pass 2 entirely (~20-30% of chunks)
- `GlinerExtractor` in `gliner_extractor.rs` is already implemented

### Pass 2: Targeted LLM (constrained, cheaper)
- Input: chunk text + entity list from Pass 1
- **New relationship-only prompt** (does not exist yet):
```
You are a relationship extractor. The following entities have been identified in the text below:

Entities:
- GraphRAG (technology)
- Leiden algorithm (algorithm)
- Microsoft Research (organization)

Given the text, extract only the relationships between these entities. Do not add new entities.

Return a JSON object with this exact schema:
{
  "relationships": [
    {
      "source_name": "source entity name (must match an entity above)",
      "target_name": "target entity name (must match an entity above)",
      "rel_type": "relationship type",
      "description": "brief description or null",
      "confidence": 0.0-1.0
    }
  ]
}
```
- Simpler task → cheaper model works (gpt-4o-mini or 3B SLM)
- Less hallucination (can't invent entities not found by GLiNER)

### 1. Add `TwoPassExtractor` in `ingestion/`
- Wraps `GlinerExtractor` + `LlmExtractor`
- Calls GLiNER first, feeds entity list into a modified LLM prompt
- Falls back to single-pass LLM if GLiNER sidecar is unavailable

### 2. Add relationship-only prompt to `LlmExtractor`
- New method: `extract_relationships(text, entities)`
- Or a second `SYSTEM_PROMPT_RELATIONSHIPS` constant
- Current `SYSTEM_PROMPT` stays for single-pass fallback

### 3. Config switch
- `COVALENCE_EXTRACTOR=llm|gliner|two_pass` env var
- Default: `llm` (current behavior, backward compatible)

### 4. GLiNER sidecar
- Python FastAPI wrapper around GLiNER2
- Default port 8432
- Docker image for easy deployment

## Benefits
- **Cost**: eliminate ~50-70% of LLM token usage
- **Quality**: grounded entities, focused relationship extraction
- **Speed**: GLiNER is local and fast; LLM calls are shorter
- **Training data**: every (chunk, entities, relationships) tuple from Pass 2 becomes training data for #11

## Dependencies
- #31 (ontology clustering) for normalizing extracted types
- GLiNER2 Python sidecar (to build)

## References
- GLiNER paper ingested (2311.08526)
- Conversation from 2026-03-09 02:20 AM established this design

**Comment by zonk1024** [comment]
Implemented in d56bbb3.

**TwoPassExtractor**: GLiNER (local NER) for entities → targeted LLM for relationships only.

- Pass 1: GLiNER sidecar extracts entities with types and confidence
- Pass 2: LLM receives entity list and extracts only relationships between them (constrained, cheaper)
- Falls back to single-pass LLM if GLiNER sidecar unavailable
- Skips LLM pass when GLiNER finds no entities (~20-30% of chunks)
- Validates LLM output: only relationships referencing known GLiNER entities are kept

**Config**: `COVALENCE_ENTITY_EXTRACTOR=two_pass` (also supports `llm` and `gliner2` as before).

**Relationship-only prompt**: New `SYSTEM_PROMPT_RELATIONSHIPS` with `extract_relationships(text, entities)` method on LlmExtractor. Simpler task = cheaper model works.
